README
CPAN::Testers::Data::Generator(3)            CPAN::Testers::Data::Generator(3)



NAME
       CPAN::Testers::Data::Generator - Download and summarize CPAN Testers
       data

SYNOPSIS
         % cpanstats
         # ... wait patiently, very patiently
         # ... then use cpanstats.db, an SQLite database
         # ... or the MySQL database

DESCRIPTION
       This distribution was originally written by Leon Brocard to download
       and summarize CPAN Testers data. However, all of the original code has
       been rewritten to use the CPAN Testers Statistics database generation
       code. This now means that all the CPAN Testers sites including the
       Reports site, the Statistics site and the CPAN Dependencies site, can
       use the same database.

       This module downloads articles from the cpan-testers newsgroup,
       generating or updating an SQLite database containing all the most
       important information. You can then query this database, or use
       CPAN::WWW::Testers to present it over the web.

       A good example query for Acme-Colour would be:

         SELECT version, status, count(*) FROM cpanstats WHERE
         distribution = "Acme-Colour" group by version, state;

       To create a database from scratch can take several days, as there are
       now over 2 million articles in the newgroup. As such updating from a
       known copy of the database is much more advisable. If you don't want to
       generate the database yourself, you can obtain the latest official copy
       (compressed with gzip) at http://devel.cpantesters.org/cpanstats.db.gz

       With over 6 million articles in the archive, if you do plan to run this
       software to generate the databases it is recommended you utilise a
       high-end processor machine. Even with a reasonable processor it can
       take a week!

SQLite DATABASE SCHEMA
       The cpanstats database schema is very straightforward, one main table
       with several index tables to speed up searches. The main table is as
       below:

         +--------------------------------+
         | cpanstats                      |
         +----------+---------------------+
         | id       | INTEGER PRIMARY KEY |
         | state    | TEXT                |
         | postdate | TEXT                |
         | tester   | TEXT                |
         | dist     | TEXT                |
         | version  | TEXT                |
         | platform | TEXT                |
         | perl     | TEXT                |
         | osname   | TEXT                |
         | osvers   | TEXT                |
         | date     | TEXT                |
         | guid     | TEXT                |
         | type     | INTEGER             |
         +----------+---------------------+

       It should be noted that 'postdate' refers to the YYYYMM formatted date,
       whereas the 'date' field refers to the YYYYMMDDhhmm formatted date and
       time.

       The metabase database schema is again very straightforward, and
       consists of one table, as below:

         +--------------------------------+
         | metabase                       |
         +----------+---------------------+
         | guid     | TEXT PRIMARY KEY    |
         | report   | TEXT                |
         +----------+---------------------+

       The report field is JSON encoded, and is a cached version of the one
       extracted from Metabase::Librarian.

SIGNIFICANT CHANGES
   v0.31 CHANGES
       With the release of v0.31, a number of changes to the codebase were
       made as a further move towards CPAN Testers 2.0. The first change is
       the name for this distribution. Now titled
       'CPAN-Testers-Data-Generator', this now fits more appropriately within
       the CPAN-Testers namespace on CPAN.

       The second significant change is to now reference a MySQL cpanstats
       database.  The SQLite version is still updated as before, as a number
       of other websites and toolsets still rely on that database file format.
       However, in order to make the CPAN Testers Reports website more
       dynamic, an SQLite database is not really appropriate for a high demand
       website.

       The database creation code is now available as a standalone program, in
       the examples directory, and all the database communication is now
       handled by the new distribution CPAN-Testers-Common-DBUtils.

   v0.41 CHANGES
       In the next stage of development of CPAN Testers 2.0, the id field used
       within the database schema above for the cpanstats table no longer
       matches the NNTP ID value, although the id in the articles does still
       reference the NNTP ID.

       In order to correctly reference the id in the articles table, you will
       need to use the function guid_to_nntp() with
       CPAN::Testers::Common::Utils, using the new guid field in the cpanstats
       table.

       As of this release the cpanstats id field is a unique auto incrementing
       field.

       The next release of this distribution will be focused on generation of
       stats using the Metabase storage API.

   v1.00 CHANGES
       Moved to Metabase API. The change to a definite major version number
       hopefully indicates that this is a major interface change. All previous
       NNTP access has been dropped and is no longer relavent. All report
       updates are now fed from the Metabase API.

INTERFACE
   The Constructor
       ·   new

           Instatiates the object CPAN::Testers::Data::Generator. Accepts a
           hash containing values to prepare the object. These are described
           as:

             my $obj = CPAN::Testers::Data::Generator->new(
                           logfile => './here/logfile',
                           config  => './here/config.ini'
             );

           Where 'logfile' is the location to write log messages. Log messages
           are only written if a logfile entry is specified, and will always
           append to any existing file. The 'config' should contain the path
           to the configuration file, used to define the database access and
           general operation settings.

   Public Methods
       ·   generate

           Starting from the last cached report, retrieves all the more recent
           reports from the Metabase Report Submission server, parsing each
           and recording each report in both the cpanstats databases (MySQL &
           SQLite) and the metabase cache database.

       ·   regenerate

           For a given date range, retrieves all the reports from the Metabase
           Report Submission server, parsing each and recording each report in
           both the cpanstats databases (MySQL & SQLite) and the metabase
           cache database.

           Note that as only 2500 can be returned at any one time due to
           Amazon SimpleDB restrictions, this method will only process the
           guids returned from a given start data, up to a maxiumu of 2500
           guids.

           This methog will return the guid of the last report processed.

       ·   rebuild

           In the event that the cpanstats database needs regenerating, either
           in part or for the whole database, this method allow you to do so.
           You may supply parameters as to the 'start' and 'end' values
           (inclusive), where all records are assumed by default. Records are
           rebuilt using the local metabase cache database.

       ·   reparse

           Rather than a complete rebuild the option to selective reparse
           selected entries is useful if there are reports which were
           previously unable to correctly supply a particular field, which now
           has supporting parsing code within the codebase.

           In addition there is the option to exclude fields from parsing
           checks, where they may be corrupted, and can be later amended using
           the 'cpanstats-update' tool.

   Private Methods
       ·   commit

           To speed up the transaction process, a commit is performed every
           500 inserts.  This method is used as part of the clean up process
           to ensure all transactions are completed.

       ·   get_next_guids

           Get the list of GUIDs for the reports that have been submitted
           since the last cached report.

       ·   already_saved

           Given a guid, determines whether it has already been saved in the
           local metabase cache.

       ·   get_fact

           Get a specific report factfor a given GUID.

       ·   parse_report

           Parses a report extracting the metadata required for the cpanstats
           database.

       ·   reparse_report

           Parses a report (from a local metabase cache) extracting the
           metadata required for the stats database.

       ·   retrieve_report

           Given a guid will attempt to return the report metadata from the
           cpanstats database.

       ·   store_report

           Inserts the components of a parsed report into the cpanstats
           database.

       ·   cache_report

           Inserts a serialised report into a local metabase cache database.

       ·   cache_update

           For the current report will update the local metabase cache with
           the id used within the cpanstats database.

HISTORY
       The CPAN testers was conceived back in May 1998 by Graham Barr and
       Chris Nandor as a way to provide multi-platform testing for modules.
       Today there are over 2 million tester reports and more than 100 testers
       each month giving valuable feedback for users and authors alike.

BECOME A TESTER
       Whether you have a common platform or a very unusual one, you can help
       by testing modules you install and submitting reports. There are plenty
       of module authors who could use test reports and helpful feedback on
       their modules and distributions.

       If you'd like to get involved, please take a look at the CPAN Testers
       Wiki, where you can learn how to install and configure one of the
       recommended smoke tools.

       For further help and advice, please subscribe to the the CPAN Testers
       discussion mailing list.

         CPAN Testers Wiki
           - http://wiki.cpantesters.org
         CPAN Testers Discuss mailing list
           - http://lists.cpan.org/showlist.cgi?name=cpan-testers-discuss

BUGS, PATCHES & FIXES
       There are no known bugs at the time of this release. However, if you
       spot a bug or are experiencing difficulties, that is not explained
       within the POD documentation, please send bug reports and patches to
       the RT Queue (see below).

       Fixes are dependent upon their severity and my availability. Should a
       fix not be forthcoming, please feel free to (politely) remind me.

       RT Queue -
       http://rt.cpan.org/Public/Dist/Display.html?Name=CPAN-Testers-Data-Generator

SEE ALSO
       CPAN::Testers::Report, Metabase, Metabase::Fact,
       CPAN::Testers::Fact::LegacyReport, CPAN::Testers::Fact::TestSummary,
       CPAN::Testers::Metabase::AWS

       CPAN::Testers::WWW::Statistics

       http://www.cpantesters.org/, http://stats.cpantesters.org/,
       http://wiki.cpantesters.org/

AUTHOR
       It should be noted that the original code for this distribution began
       life under another name. The original distribution generated data for
       the original CPAN Testers website. However, in 2008 the code was
       reworked to generate data in the format for the statistics data
       analysis, which in turn was reworked to drive the redesign of the all
       the CPAN Testers websites. To reflect the code changes, a new name was
       given to the distribution.

   CPAN-WWW-Testers-Generator
         Original author:    Leon Brocard <acme@astray.com>   (C) 2002-2008
         Current maintainer: Barbie       <barbie@cpan.org>   (C) 2008-2010

   CPAN-Testers-Data-Generator
         Original author:    Barbie       <barbie@cpan.org>   (C) 2008-2011

LICENSE
       This code is distributed under the Artistic License 2.0.



perl v5.10.1                      2011-07-04 CPAN::Testers::Data::Generator(3)


CHANGES
Revision history for Perl module CPAN::Testers::Data::Generator.

        - script updates:
            bin/cpanstats                       (v1.04)

1.03    2012-10-21
        - spelling fixes. (thanks to Florian Schlichting & Gregor Herrmann).
        - AWS values are now configure settings, that can be overridden on the
          command line.
        - compensate for Amazon SimpleDB's brokeness with dates.
        - added parse function for specific GUIDs.
        - change file dates changed to meet W3CDTF standards.
        - META tests no longer require specific versions.
        - reworked Makefile.PL for clarity.
        - License updated to Artistic License v2.0.
        - added minimum perl version (5.006).
        - added missing prerequisite.

1.02    2012-04-01
        - reworked regenerate functionality and script.
        - reworked reparse functionality and script to use Metabase API.
        - added handling of localonly and check options.
        - updated script:
            bin/cpanstats                       (v1.02)

1.01    2012-01-22
        - generator now builds a JSON and RSS file of the last n reports.
        - added id to page_request insert (to speed up page builder).
        - reworked regenerator to allow a file of guid/date ranges.
        - update regenerate option and documentation:
            bin/cpanstats                       (v1.01)
        - added mappings for old badly formatted distributions.

1.00    2010-11-14
        - redesigned to use Metabase. All NNTP references removed.
        - removed ignore and nostore options.
        - added nonstop option.
        - added error report mail alert, to highlight problems quickly
        - improved validation of metabase results, due to the search sometimes
          returning older reports!
        - improved retrieval of tester fact
        - scripts removed:
            bin/articles-verify
            bin/articles-view
            bin/cpanstats-reparse
            bin/cpanstats-verify
            examples/cpanstats-convert
        - added guid command line options and/or output:
            bin/cpanstats-select                (v0.11)
            bin/cpanstats-update                (v0.09)
            bin/cpanstats-delete                (v0.14)
        - major update of all SQL tables and databases
            examples/cpanstats-createdb         (v0.02)
        - added META.json
        - added regenerate() and retrieve_report() methods.
        - added regenerate option and documentation:
            bin/cpanstats                       (v1.00)
        - various data amendments, including _check_arch_os(), as the metabase
          uses the raw data.
        - new perl_version table incorporated to speed up perl version lookups.

0.41    2010-03-18
        - fixes to change the 'id' (was NNTP ID) to an auto incremental field.
        - reworked logic to better fit latest changes.
        - added repository to META.yml.
        - documentation updates.

0.40    2010-02-02
        - fixes to accommodate GUID changes.
        - added support for 'type' field.

0.39    2010-01-30
        - major update to include guid for Metabase/CT2.0.

0.38    2009-10-19
        - added functionality to update the osname table.
        - added cpanstats DB schema to examples directory.

0.37    2009-06-07
        - added page request functionality.
        - added release summary functionality.
        - due to the recent expansion, the 12generate.t test script now needs
          more complex logic. The intention is to use Test::Database for this,
          but unfortunately T::D isn't stable enough to use yet. As such this
          test script is now effectively disabled.

0.36    2009-02-16
        - clean up of pre-requisites.
        - CPAN::Testers::Data::Generator::Article abstracted into its own
          distribution CPAN-Testers-Common-Article.
        - fixed scripts due to above:
            bin/cpanstats         (v0.07)
            bin/cpanstats-reparse (v0.10)

0.35    2009-01-29
        - updated patterns when extracting perl version (Article.pm).
        - updated patterns when extracting osname (Article.pm).
        - updated cannot connect to NNTP server error message.

0.34    2009-01-12
        - updated Copyright info, minor fixes and added 'Getopt::ArgvFile', if
          not already used:
            bin/articles-verify   (v0.03)
            bin/articles-view     (v0.03)
            bin/cpanstats         (v0.06)
            bin/cpanstats-delete  (v0.13)
            bin/cpanstats-reparse (v0.09)
            bin/cpanstats-select  (v0.10)
            bin/cpanstats-update  (v0.08)
            bin/cpanstats-verify  (v0.24)
        - added cpanstats-verify.sh to examples.
        - added missing log file command line option to bin/cpanstats (v0.06).
        - added missing prerequisites (thanks to CPANTS)

0.33    2008-12-29
        - updated help text to bin/cpanstats-verify  (v0.23)
        - fix to stop existing records being reinserted.

0.32    2008-12-19
        - added progress option to bin/cpanstats-verify (v0.22)
        - fixes to Article.pm to attempt to correct missing osnames and osvers.
        - added articles scripts that have been used behind the scenes for
          sometime:
            bin/articles-verify   (v0.02)
            bin/articles-view     (v0.02)

0.31    2008-12-18
        - name change to fit below the CPAN::Testers namespace.
        - added epoch to date extraction from article.
        - added filename reference from CPAN-DistnameInfo.
        - added ability to provide SQL string in bin/cpanstats-select.
        - added better command line option parsing to bin/cpanstats-reparse.
        - convert to use CPAN::Testers::Common::DBUtils.
        - added MySQL cpanstats database to codebase.
        - new scripts added to examples to handle database creation:
            examples/cpanstats-createdb (v0.01)
            examples/cpanstats-convert  (v0.01)

lib/CPAN/Testers/Data/Generator/Article.pm
package CPAN::Testers::Data::Generator::Article;

use warnings;
use strict;
use vars qw($VERSION);

$VERSION = '0.35';

#----------------------------------------------------------------------------
# Library Modules

use CPAN::DistnameInfo;
use Email::Simple;
use MIME::Base64;
use MIME::QuotedPrint;
use Time::Local;

use base qw( Class::Accessor::Fast );

#----------------------------------------------------------------------------
# Variables

my %month = (
	Jan => 1, Feb => 2, Mar => 3, Apr => 4,  May => 5,  Jun => 6,
	Jul => 7, Aug => 8, Sep => 9, Oct => 10, Nov => 11, Dec => 12,
);

my %regexes = (
    # with time
    1 => { re => qr/(?:\w+,)?\s+(\d+)\s+(\w+)\s+(\d+)\s+(\d+):(\d+)/,   f => [qw(day month year hour min)] },     # Wed, 13 September 2004 06:29
    2 => { re => qr/(\d+)\s+(\w+)\s+(\d+)\s+(\d+):(\d+)/,               f => [qw(day month year hour min)] },     # 13 September 2004 06:29
    3 => { re => qr/(\w+)?\s+(\d+),?\s+(\d+)\s+(\d+):(\d+)/,            f => [qw(month day year hour min)] },     # September 22, 1999 06:29

    # just the date
    4 => { re => qr/(?:\w+,)?\s+(\d+)\s+(\w+)\s+(\d+)/, f => [qw(day month year)] },  # Wed, 13 September 2004
    5 => { re => qr/(\d+)\s+(\w+)\s+(\d+)/,             f => [qw(day month year)] },  # 13 September 2004
    6 => { re => qr/(\w+)?\s+(\d+),?\s+(\d+)/,          f => [qw(month day year)] },  # September 22, 1999 06:29
);

my $OSNAMES = qr/(cygwin|freebsd|netbsd|openbsd|darwin|linux|cygwin|darwin|MSWin32|dragonfly|solaris|MacOS|irix|mirbsd|gnu|bsdos|aix|sco|os2)/i;
my %OSNAMES = (
    'MacPPC'    => 'macos',
    'osf'       => 'dec_osf',
    'pa-risc'   => 'hpux',
    's390'      => 'os390',
    'VMS_'      => 'vms',
    'ARCHREV_0' => 'hpux',
);

#----------------------------------------------------------------------------
# The Application Programming Interface

__PACKAGE__->mk_accessors(qw(
                    postdate date epoch status from distribution version
                    perl osname osvers archname subject author filename));

sub new {
    my($class, $article) = @_;
    my $self = {};
    bless $self, $class;

    $article = decode_qp($article)	if($article =~ /=3D/);

    my $mail = Email::Simple->new($article);
    return unless $mail;
    return if $mail->header("In-Reply-To");

    my $from    = $mail->header("From");
    my $subject = $mail->header("Subject");
    return unless $subject;
    return if $subject =~ /::/; # it's supposed to be a distribution

    $self->{mail}    = $mail;
    $self->{from}    = $from;
    $self->{subject} = $subject;

    ($self->{postdate},$self->{date},$self->{epoch}) = _parse_date($mail);

    return $self;
}

sub _parse_date {
    my $mail = shift;
    my ($date1,$date2,$date3) = _extract_date($mail->header("Date"));
    my @received  = $mail->header("Received");

    for my $hdr (@received) {
        next    unless($hdr =~ /.*;\s+(.*)\s*$/);
        my ($dt1,$dt2,$dt3) = _extract_date($1);
        if($dt2 > $date2 + 1200) {
            $date1 = $dt1;
            $date2 = $dt2;
            $date3 = $dt3;
        }
    }

#print STDERR "        ... X.[Date: ".($date||'')."]\n";
    return($date1,$date2,$date3);
}

sub _extract_date {
    my $date = shift;
    my (%fields,@fields,$index);

#print STDERR "#        ... 0.[Date: ".($date||'')."]\n";

    for my $inx (sort {$a <=> $b} keys %regexes) {
        (@fields) = ($date =~ $regexes{$inx}->{re});
        if(@fields) {
            $index = $inx;
            last;
        }
    }

    return('000000','000000000000',0) unless(@fields && $index);

    @fields{@{$regexes{$index}->{f}}} = @fields;

    $fields{month} = substr($fields{month},0,3);
    $fields{mon}   = $month{$fields{month}};
    return('000000','000000000000',0) unless($fields{mon} && $fields{year} > 1998);

    $fields{$_} ||= 0          for(qw(sec min hour day mon year));
    my @date = map { $fields{$_} } qw(sec min hour day mon year);

#print STDERR "#        ... 1.[$_][$fields{$_}]\n"   for(qw(year month day hour min));
    my $short = sprintf "%04d%02d",             $fields{year}, $fields{mon};
    my $long  = sprintf "%04d%02d%02d%02d%02d", $fields{year}, $fields{mon}, $fields{day}, $fields{hour}, $fields{min};
    $date[4]--;
    my $epoch = timelocal(@date);

    return($short,$long,$epoch);
}

sub parse_upload {
    my $self = shift;
    my $mail = $self->{mail};
    my $subject = $self->{subject};

    return 0	unless($subject =~ /CPAN Upload:\s+([-\w\/\.\+]+)/i);
    my $distvers = $1;

    # only record supported archives
    return 0    if($distvers !~ /\.(?:(?:tar\.|t)(?:gz|bz2)|zip)$/);

    # CPAN::DistnameInfo doesn't support .tar.bz2 files ... yet
    $distvers =~ s/\.(?:tar\.|t)bz2$//i;
    $distvers .= '.tar.gz' unless $distvers =~ /\.(?:(?:tar\.|t)gz|zip)$/i;

    # CPAN::DistnameInfo doesn't support old form of uploads
    my @parts = split("/",$distvers);
    if(@parts == 2) {
        my ($first,$second,$rest) = split(//,$distvers,3);
        $distvers = "$first/$first$second/$first$second$rest";
    }

    my $d = CPAN::DistnameInfo->new($distvers);
    $self->distribution($d->dist);
    $self->version($d->version);
    $self->author($d->cpanid);
    $self->filename($d->filename);

    return 1;
}

sub parse_report {
    my $self = shift;
    my $mail = $self->{mail};
    my $from = $self->{from};
    my $subject = $self->{subject};

    my ($status, $distversion, $platform, $osver) = split /\s+/, $subject;
    return 0  unless $status =~ /^(PASS|FAIL|UNKNOWN|NA)$/i;

    $platform ||= "";
    $platform =~ s/[\s&,<].*//;

    $distversion =~ s!/$!!;
    $distversion =~ s/\.tar.*/.tar.gz/;
    $distversion .= '.tar.gz' unless $distversion =~ /\.(tar|tgz|zip)/;

    my $d = CPAN::DistnameInfo->new($distversion);
    my ($dist, $version) = ($d->dist, $d->version);
    return 0 unless defined $dist;
    return 0 unless defined $version;

    my $encoding = $mail->header('Content-Transfer-Encoding');

    my $body = $mail->body;
    $body = decode_base64($body)  if($encoding && $encoding eq 'base64');

    my $perl = $self->_extract_perl_version(\$body);

    my ($osname)   = $body =~ /(?:Summary of my perl5|Platform:).*?osname=([^\s\n,<\']+)/s;
    my ($osvers)   = $body =~ /(?:Summary of my perl5|Platform:).*?osvers=([^\s\n,<\']+)/s;
    my ($archname) = $body =~ /(?:Summary of my perl5|Platform:).*?archname=([^\s\n&,<\']+)/s;
    $archname =~ s/\n.*//	if($archname);

    $self->status($status);
    $self->distribution($dist);
    $self->version($version);
    $self->from($from || "");
    $self->perl($perl);
    $self->filename($d->filename);

    unless($archname || $platform) {
  	    if($osname && $osvers)	{ $platform = "$osname-$osvers" }
	    elsif($osname)		    { $platform = $osname }
    }

    unless($osname) {
        for my $text ($platform, $archname) {
            next    unless($text);
            if($text =~ $OSNAMES) {
                $osname = $1;
            } else {
                for my $rx (keys %OSNAMES) {
                    if($text =~ /$rx/i) {
                        $osname = $OSNAMES{$rx};
                        last;
                    }
                }
            }
            last    if($osname);
        }
    }

    $osvers ||= $osver;

    $self->osname($osname || "");
    $self->osvers($osvers || "");
    $self->archname($archname || $platform);

    return 1;
}

sub passed {
    my $self = shift;
    return $self->status eq 'PASS';
}

sub failed {
    my $self = shift;
    return $self->status eq 'FAIL';
}

# there are a few old test reports that omitted the perl version number.
# In these instances 0 is assumed. These reports are now so old, that
# worrying about them is not worth the effort.

sub _extract_perl_version {
    my ($self, $body) = @_;

    # Summary of my perl5 (revision 5.0 version 6 subversion 1) configuration:
    # Summary of my perl5 (revision 5 version 10 subversion 0) configuration:
    my ($rev, $ver, $sub, $extra) =
        $$body =~ /Summary of my (?:perl(?:\d+)?)? \((?:revision )?(\d+(?:\.\d+)?) (?:version|patchlevel) (\d+) subversion\s+(\d+) ?(.*?)\) configuration/si;

    if(defined $rev) {
        my $perl = $rev + ($ver / 1000) + ($sub / 1000000);
        $rev = int($perl);
        $ver = int(($perl*1000)%1000);
        $sub = int(($perl*1000000)%1000);

        my $version = sprintf "%d.%d.%d", $rev, $ver, $sub;
        $version .= " $extra" if $extra;
        return $version;
    #   return sprintf "%0.6f", $perl;	# an alternate format
    }

    # the following is experimental and may provide incorrect data

    ($rev, $ver, $sub) =
        $$body =~ m!/(?:(?:site_perl|perl|perl5|\.?cpanplus)/|perl-)(5)\.?([6-9]|1[0-2])\.?(\d+)/!;
    if(defined $rev) {
        my $version = sprintf "%d.%d.%d", $rev, $ver, $sub;
        return $version;
    }


#    warn "Cannot parse perl version for article:\n$body";
    return 0;
}

1;

__END__

=head1 NAME

CPAN::Testers::Data::Generator::Article - Parse a CPAN Testers article

=head1 DESCRIPTION

This is used by CPAN::Testers::Data::Generator.

=head1 INTERFACE

=head2 The Constructor

=over 4

=item * new

The constructor. Pass in a reference to the article.

=back

=head2 Methods

=over 4

=item * parse_upload

Parses an upload article.

=item * parse_report

Parses a report article.

=item * passed

Whether the report was a PASS

=item * failed

Whether the report was a FAIL

=back

=head2 Accessors

All the following are accessors available through via the object, once an
article has been parsed as a report or upload announcement.

=over 4

=item * postdate

'YYYYMM' representation of the date article was posted.

=item * date

'YYYYMMDDhhmm' representation of the date article was posted.

=item * status

For reports this will be the grade, for uploads this will be 'CPAN'.

=item * from

Who posted the article.

=item * distribution

The distribution name.

=item * version

The distribution version.

=item * perl

The perl interpreter version used for testing.

=item * osname

Operating system name.

=item * osvers

Operating system version.

=item * archname

Operating system architecture name. This is usually based on the osname and
osvers, but they are not always the same.

=item * subject

Subject line of the original post.

=back

=head1 BUGS, PATCHES & FIXES

There are no known bugs at the time of this release. However, if you spot a
bug or are experiencing difficulties, that is not explained within the POD
documentation, please send bug reports and patches to the RT Queue (see below).

Fixes are dependant upon their severity and my availablity. Should a fix not
be forthcoming, please feel free to (politely) remind me.

RT Queue -
http://rt.cpan.org/Public/Dist/Display.html?Name=CPAN-Testers-Data-Generator

=head1 SEE ALSO

L<CPAN::WWW::Testers>,
L<CPAN::Testers::WWW::Statistics>

F<http://www.cpantesters.org/>,
F<http://stats.cpantesters.org/>,
F<http://wiki.cpantesters.org/>

=head1 AUTHOR

  Original author:    Leon Brocard <acme@astray.com>   (C) 2002-2008
  Current maintainer: Barbie       <barbie@cpan.org>   (C) 2008-2009

=head1 LICENSE

This code is distributed under the same license as Perl.


Makefile.PL
#!/usr/bin/perl

use strict;
use warnings;
use 5.006;

use ExtUtils::MakeMaker;

WriteMakefile(
    AUTHOR              => 'Barbie <barbie@cpan.org>',
    NAME                => 'CPAN::Testers::Data::Generator',
    VERSION_FROM        => 'lib/CPAN/Testers/Data/Generator.pm',
    ABSTRACT            => 'CPAN Testers cpanstats Database Generator',
    NO_META             => 1,

    PREREQ_PM           => {

        # prereqs
        'Config::IniFiles'                  => 0,
        'CPAN::Testers::Common::Article'    => 0.36,
        'CPAN::Testers::Common::DBUtils'    => 0.03,
        'CPAN::Testers::Fact::LegacyReport' => 0,
        'CPAN::Testers::Fact::TestSummary'  => 0,
        'CPAN::Testers::Metabase::AWS'      => 0,
        'CPAN::Testers::Report'             => 0,
        'File::Basename'                    => 0,
        'File::Path'                        => 0,
        'File::Slurp'                       => 0,
        'HTML::Entities'                    => 0,
        'IO::File'                          => 0,
        'JSON'                              => 0,
        'Metabase'                          => 0.014,
        'Metabase::Fact'                    => 0,
        'Time::Local'                       => 0,
        'XML::Feed'                         => 0,

        'DBI'                               => 0,
        'DBD::mysql'                        => 0,
        'DBD::SQLite'                       => 1.07,

        # used by run scripts
        'Getopt::Long'                      => 0,
        'Getopt::ArgvFile'                  => 0,

        # build/test prereqs
        'Test::More'                        => '0.70'

    },

    EXE_FILES           => [ 'bin/cpanstats' ]
);
